### ISSUE TITLE
Devise helpers not available

### ISSUE THREAD

**Issue by wolak88** [graphql-devise/graphql_devise#151]
### Describe the bug

My controllers don't have access to devise helpers in new rails 6 application in API mode.

### Environment

```
source "https://rubygems.org"
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby "3.1.0"

gem "rails", "~> 6.1.1"
gem "pg", "~> 1.1"
gem "puma", "~> 5.0"
gem "bootsnap", ">= 1.4.4", require: false
gem "rack-cors"

gem 'graphql'
gem 'graphql_devise'

gem "administrate"
gem 'acts_as_tenant'

group :development, :test do
  gem "annotate"
  gem "byebug", platforms: %i[mri mingw x64_mingw]
end

group :development do
  gem "listen", "~> 3.3"
  gem "rubocop", require: false
  gem "rubocop-performance", require: false
  gem "rubocop-rails", require: false
  gem "rubocop-rspec", require: false
  gem "graphiql-rails"
  gem "letter_opener"
end

gem "tzinfo-data", platforms: %i[mingw mswin x64_mingw jruby]
```

Application controller:

```
class ApplicationController < ActionController::API
  include GraphqlDevise::Concerns::SetUserByToken
  include SetCurrentRequestDetails
end
```

SetCurrentRequestDetails:

```
module SetCurrentRequestDetails
  extend ActiveSupport::Concern

  included do |base|
    if base < ActionController::API
      set_current_tenant_through_filter
      before_action :set_request_details
    end
  end

  def set_request_details
    Current.request_id = request.uuid
    Current.user_agent = request.user_agent
    Current.ip_address = request.ip
    Current.user = current_user - LINE WHICH CAUSES ERROR

    Current.account ||= account_from_domain || account_from_subdomain || account_from_session || fallback_account

    set_current_tenant(Current.account)
  end
end

```

I've tried to add 
 `before_action :authenticate_user!`
but its also not available

### Steps to reproduce

1. Run any query/mutation

### Expected behavior

current_user should be nil or user from token

### Actual behavior

```
Error:
Started POST "/graphql" for ::1 at 2021-02-16 17:34:19 +0100
Processing by GraphqlController#execute as */*
  Parameters: {"query"=>"# mutation userSignUp {\n#     userSignUp(email: \"wolak88@gmail.com\",\n#       password: \"password\",\n#       passwordConfirmation: \"password\",\n#       confirmSuccessUrl: \"localhost:3000/\")\n#     {\n#       user {\n#         email\n#       }\n#     }\n#   }\n\nmutation {\n  userSignUp(\n    email: \"wolak88+a@gmail.com\",\n      password: \"password\",\n      passwordConfirmation: \"password\",\n      confirmSuccessUrl: \"localhost:3000/\"\n  ) {\n    credentials { accessToken }\n    authenticatable {\n      email\n    }\n  }\n}", "variables"=>nil, "graphql"=>{"query"=>"# mutation userSignUp {\n#     userSignUp(email: \"random@gmail.com\",\n#       password: \"password\",\n#       passwordConfirmation: \"password\",\n#       confirmSuccessUrl: \"localhost:3000/\")\n#     {\n#       user {\n#         email\n#       }\n#     }\n#   }\n\nmutation {\n  userSignUp(\n    email: \"wolak88+a@gmail.com\",\n      password: \"password\",\n      passwordConfirmation: \"password\",\n      confirmSuccessUrl: \"localhost:3000/\"\n  ) {\n    credentials { accessToken }\n    authenticatable {\n      email\n    }\n  }\n}", "variables"=>nil}}
Completed 500 Internal Server Error in 5ms (Allocations: 2105)


  
NameError (undefined local variable or method `current_user' for #<GraphqlController:0x0000000000a690>
Did you mean?  current_tenant):
  
app/controllers/set_current_request_details.rb:17:in `set_request_details'
```

**Comment by mcelicalderon** [comment]
Hey @wolak88! I will look into this ASAP. In the mean time, could you please share you Graphql controller's code? I think the problem might be there. Also your application's controller would be useful if you didn't already share a full version of that one.
Another question, so you are going to be setting current_user before actually getting to the GQL schema. Are you using a separate route for the auth schema or using the schema plugin?

**Comment by wolak88** [comment]
Yes, I've shared all of the contents of my `ApplicationController` as for gql controller it's also a generic one:
```
class GraphqlController < ApplicationController
  def execute
    variables = prepare_variables(params[:variables])
    query = params[:query]
    operation_name = params[:operationName]
    result = TradeHubSchema.execute(query, variables: variables, context: graphql_context(:user), operation_name: operation_name)
    render json: result unless performed?
  rescue => e
    raise e unless Rails.env.development?
    handle_error_in_development e
  end

  private

  # Handle variables in form data, JSON body, or a blank value
  def prepare_variables(variables_param)
    case variables_param
    when String
      if variables_param.present?
        JSON.parse(variables_param) || {}
      else
        {}
      end
    when Hash
      variables_param
    when ActionController::Parameters
      variables_param.to_unsafe_hash # GraphQL-Ruby will validate name and type of incoming variables.
    when nil
      {}
    else
      raise ArgumentError, "Unexpected parameter: #{variables_param}"
    end
  end

  def handle_error_in_development(e)
    logger.error e.message
    logger.error e.backtrace.join("\n")

    render json: { errors: [{ message: e.message, backtrace: e.backtrace }], data: {} }, status: 500
  end
end
```

I'm using it as a plugin, here's my schema (also generic one):

```
class TradeHubSchema < GraphQL::Schema
  use GraphqlDevise::SchemaPlugin.new(
    query: Types::QueryType,
    mutation: Types::MutationType,
    resource_loaders: [
      GraphqlDevise::ResourceLoader.new("User"),
    ]
  )
  mutation(Types::MutationType)
  query(Types::QueryType)

  # Union and Interface Resolution
  def self.resolve_type(abstract_type, obj, ctx)
    # TODO: Implement this function
    # to return the correct object type for `obj`
    raise(GraphQL::RequiredImplementationMissingError)
  end

  # Relay-style Object Identification:

  # Return a string UUID for `object`
  def self.id_from_object(object, type_definition, query_ctx)
    # Here's a simple implementation which:
    # - joins the type name & object.id
    # - encodes it with base64:
    # GraphQL::Schema::UniqueWithinType.encode(type_definition.name, object.id)
  end

  # Given a string UUID, find the object
  def self.object_from_id(id, query_ctx)
    # For example, to decode the UUIDs generated above:
    # type_name, item_id = GraphQL::Schema::UniqueWithinType.decode(id)
    #
    # Then, based on `type_name` and `id`
    # find an object in your application
    # ...
  end
end
```


Yes I need to set it before as I'm using `act_as_tenant` gem for row level multitenancy, My users can have multiple accounts depending on domain that the request is coming from so I need to set current tenant before going into graphQL controller.

**Comment by mcelicalderon** [comment]
Thank you for that. Ok, I think I know what's going on. When using the plugin, we defer setting the user to the GQL schema, precisely because when eager loading is disabled, plugin configuration will only run the first time a query is executed on the schema and Devise  needs this configuration in order to enable methods like `current_user` in the controller. This is something we are going to fix in the next big release so we don't depend on Devise's model configuration. In the mean time here are a couple of alternatives that I think should work:
```
1. Enable eager-load on dev environments (not a huge fan but at least will confirm what I'm saying)
2. Use a separate route to load the schema (the old way, as described in the readme file). You can even do this as a hack and have the plugin in the schema too. Putting the gem's config in the routes file will make sure your model is configured before making the first request thus making those dynamic methods available in the controller.
3. Make sure via an initializer or another method your GQL schema is loaded before the first request is made. This will make the plugin code execute and configure your User model properly before the first request is executed.
```
Let me know if that worked, and as I said before, we are aware of this problem and working on a better fix soon. The new implementation should fix a couple of issues.

**Comment by wolak88** [comment]
Thanks for so quick response. Approach 1 and 2 are working, I don't how to load schema in initializer for 3rd one to work. I think the 3rd one seems the best as in the first one I would have to enable eager_load on the test environment as well and in the second one it's polluting production routes. Do you have any suggestions?

Also, I have one more question I have to monkey patch some of the mutations to support multiple accounts and I'm doing it like this do you think it's a good approach:
Initializer:
```
module Extend
  module SignUp
    def resolve(confirm_success_url: nil, **attrs)
      MonkeyPatch here
    end
  end
end

Rails.application.config.to_prepare do
  GraphqlDevise::Mutations::SignUp.class_eval do
    prepend Extend
  end
end
```

And one more, my introspection query is not working and I need that to dynamically generate types for TypeScript is it intended behaviour so only authenticated users can introspect?

**Comment by mcelicalderon** [comment]
For the third one you could probably just create an initializer like this:
```ruby
# config/initializers/graphql_devise.rb
TradeHubSchema.graphql_definition
```
About monkey patching, I always advise against that. Maybe overriding the default mutation is enough as it will allow you to override default behavior for the signUp mutation. Look at [this](https://github.com/graphql-devise/graphql_devise#available-mount-options), specifically the `operations` option.

**Comment by wolak88** [comment]
Thank you very much, I got it working with:
```
Rails.application.config.after_initialize do
  MySchema
end
```

I had one more question that you've probably missed about introspection query, why it's unavailable?

**Comment by wolak88** [comment]
@mcelicalderon 
Is it possible to make introspection query available when using `graphql_devise`?
In my previous projects I was using `disable_introspection_entry_points if Rails.env.something?` but now I'm not able to introspect at all;/

**Comment by mcelicalderon** [comment]
Hey @wolak88! Yes, this gem doesn't mess with the introspection query. How are you seeing this error? Are you getting an authentication error when using the instrospection query? If because of your multitenant gem you authenticate the user before actually getting to the GQL schema, the request is going to fail before actually trying to execute the introspection query unless you send authentication headers on the request. Just guessing here, but is this related to the graphiql gem? What I usually do is enable a second endpoint on non production environments and also make the schema available there with no authentication. That way the graphiql gem can me mounted using that endpoint instead of the authenticated one. Another alternative is to use a browser extension like mod headers to send authentication headers when using the graphiql tool. Let me know if that makes sense or how you are trying to use the introspection query.

**Comment by wolak88** [comment]
@mcelicalderon Ok, I've found the problem, by default all root fields have authenticatication enabled (`authenticate_default` on mounting the plugin) it also includes schema field! After changing it to false my introspection query started to work. Thanks for pointing me in the right direction. 

I'm just a bit worried about the consequences of that change: should any graphql-devise fields have authentication enabled? Is there a way to easily disable authentication on schema field?

**Comment by mcelicalderon** [comment]
You are right about that. Currently there's no way to disable authentication on the schema field, but we should definitively enable a way to do that. This issue already covers too much, could you please open another one about this as a feature request? I'm closing this one as all other concerns have been resolved and will be considered for the next version. Thanks!

**Comment by mcelicalderon** [comment]
The latest release of the gem fixes this. Still not documented on the Readme file, but you can now pass an argument to the plugin's initializer (last one) which will allow public introspection if set to `true`. Defaults to `!Rails.env.production?`

**Comment by mcelicalderon** [comment]
@wolak88 we just merged #175 and that should allow you to remove the after_initialize code. Now authentication should always happen in the controller and just pass current_reource to the gql_context. This will be released soon in `v0.15.0`

**Comment by wolak88** [comment]
@mcelicalderon Thanks for keeping me updated, great work! Your work is much appreciated!
